### DISCUSSION TITLE
Known use cases for Overlays

### DISCUSSION THREAD

**Discussion by ponelat** [OAI/Overlay-Specification#9] [General]
As we shape Overlays, it'll be most helpful to target specific use cases that prompted this specification in the first place.

**Comment by ponelat** [comment]
A quick first pass at OpenAPI issues that _might_ be helped with Overlays and a crude categorization of them...

### Categories

- I18n / Multilingual support for OpenAPI
- Traits/Groups of re-usable components (eg: parameters/responses). 
- Envelopes/placeholders (ie: pagination)
- Context specific API definitions (ie: one API definition, different views for different audiences)

### Issues

- https://github.com/OAI/OpenAPI-Specification/issues/2281
- https://github.com/OAI/OpenAPI-Specification/issues/1843
- https://github.com/OAI/OpenAPI-Specification/issues/1722
- https://github.com/OAI/OpenAPI-Specification/issues/1442
- https://github.com/OAI/OpenAPI-Specification/issues/1364
- https://github.com/OAI/OpenAPI-Specification/issues/445
- https://github.com/OAI/OpenAPI-Specification/issues/613
- https://github.com/OAI/OpenAPI-Specification/issues/2201
- https://github.com/OAI/OpenAPI-Specification/issues/1497
- https://github.com/OAI/OpenAPI-Specification/issues/2298
- https://github.com/OAI/OpenAPI-Specification/issues/2026
- https://github.com/OAI/OpenAPI-Specification/issues/1453
- https://github.com/OAI/OpenAPI-Specification/issues/1853
- https://github.com/OAI/OpenAPI-Specification/issues/1749
- https://github.com/OAI/OpenAPI-Specification/issues/773
- https://github.com/OAI/OpenAPI-Specification/issues/555
- https://github.com/OAI/OpenAPI-Specification/issues/1740
- https://github.com/OAI/OpenAPI-Specification/issues/1574
- https://github.com/OAI/OpenAPI-Specification/issues/1533
- https://github.com/OAI/OpenAPI-Specification/issues/1339
- https://github.com/OAI/OpenAPI-Specification/issues/1497
- https://github.com/OAI/OpenAPI-Specification/issues/1462
- https://github.com/OAI/OpenAPI-Specification/issues/756
- https://github.com/OAI/OpenAPI-Specification/issues/569
- https://github.com/OAI/OpenAPI-Specification/issues/563
- https://github.com/OAI/OpenAPI-Specification/issues/560
- https://github.com/OAI/OpenAPI-Specification/issues/519
- https://github.com/OAI/OpenAPI-Specification/issues/433
- https://github.com/OAI/OpenAPI-Specification/issues/417
- https://github.com/OAI/OpenAPI-Specification/issues/398
- https://github.com/OAI/OpenAPI-Specification/issues/369
- https://github.com/OAI/OpenAPI-Specification/issues/298
- https://github.com/OAI/OpenAPI-Specification/issues/213

**Comment by earth2marsh** [comment]
Copying my comments from the [Feb 15, 2022 discussion](https://github.com/OAI/Overlay-Specification/discussions/16#discussioncomment-2182709):

> Another use case worth noting is how middleware might transform an API surface. For example, the moment you proxy an API, you've changed the server. You might change the auth model, say from mtls to OAuth. There are many changes like this I can imagine, but I think that this kind of transformation is worth adding as a use case.
> 
> Another is to add additional metadata for codegen or test generation. An overlay might help to keep those separate. Continuing on that line of thought... examples can really blow up the size of an OpenAPI document. Why not maintain examples separately and mix them in with an overlay (which brings up the question of what approaches would feel relatively more natural for authors).
> 
> One last thought... from an end user perspective it is best to have a flattened version of a multi-part document (for example, no load failure from unavailable resource or permission prob). If we can make that assumption, it may help simplify how we think about the authoring process.

**Comment by kevinswiber** [comment]
With respect to the envelope/placeholder use case, there are some options with JSON Schema alone.  See here: https://github.com/OAI/Overlay-Specification/discussions/20#discussioncomment-2376571

**Comment by kscheirer** [comment]
- https://github.com/OAI/OpenAPI-Specification/issues/1740, added [example solutions](https://github.com/OAI/OpenAPI-Specification/issues/1740#issuecomment-1122842200) there, overlay seems good
- https://github.com/OAI/OpenAPI-Specification/issues/563, added [example solutions](https://github.com/OAI/OpenAPI-Specification/issues/563#issuecomment-1132092013) there, overlay seems good
- https://github.com/OAI/OpenAPI-Specification/issues/433, added [example solution](https://github.com/OAI/OpenAPI-Specification/issues/433#issuecomment-1132122476) there, overlay seems good

**Comment by kscheirer** [comment]
- https://github.com/OAI/OpenAPI-Specification/issues/1574, added [example solution](https://github.com/OAI/OpenAPI-Specification/issues/1574#issuecomment-1133368615), overlay works, but using component.headers in both request parameter and response header is awkward
- https://github.com/OAI/OpenAPI-Specification/issues/1339, added [example solution](https://github.com/OAI/OpenAPI-Specification/issues/1339#issuecomment-1133394397), overlay seems good

**Comment by kscheirer** [comment]
- https://github.com/OAI/OpenAPI-Specification/issues/569, visibility same as https://github.com/OAI/OpenAPI-Specification/issues/433, overlay does not help with ordering of operations within spec
- https://github.com/OAI/OpenAPI-Specification/issues/1853, overlay + traits seem good, no examples needed

**Comment by kscheirer** [comment]
- https://github.com/OAI/OpenAPI-Specification/issues/445, added [example solutions](https://github.com/OAI/OpenAPI-Specification/issues/445#issuecomment-1140068908), overlay seems good
- https://github.com/OAI/OpenAPI-Specification/issues/1364, comment and link to #445

I'm just going to update this comment to keep track of which ones I've responded on. I jumped around and picked from all over to start with.

Overlays could be helpful (10/33):
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/2281
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/1843
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/1722
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/1442
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1364
- [x] https://github.com/OAI/OpenAPI-Specification/issues/445
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/613
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/2201
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/1497
- [x] https://github.com/OAI/OpenAPI-Specification/issues/2026
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1853
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1740
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1574
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1339
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/1497
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/1462
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/756
- [x] https://github.com/OAI/OpenAPI-Specification/issues/569
- [x] https://github.com/OAI/OpenAPI-Specification/issues/563
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/560
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/519
- [x] https://github.com/OAI/OpenAPI-Specification/issues/433
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/417
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/398
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/369
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/298
- [ ] https://github.com/OAI/OpenAPI-Specification/issues/213

Overlays do not look helpful (6/33):
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1749
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1453
- [x] https://github.com/OAI/OpenAPI-Specification/issues/1533
- [x] https://github.com/OAI/OpenAPI-Specification/issues/2298
- [x] https://github.com/OAI/OpenAPI-Specification/issues/773
- [x] https://github.com/OAI/OpenAPI-Specification/issues/555

**Comment by ponelat** [comment]
Thank you for the work going through the examples, @kscheirer !

**Reply by kscheirer** [reply]
thanks @ponelat, just updated a few more! sorry for the delay, my august became very busy.

**Comment by rafalkrupinski** [comment]
Just wanted to add my use cases as a client generator developer:

- 3rd party description updates:

    I'm working with a OpenAPI description that doesn't 100% represent its service. E.g. some properties are missing nullable attribute making the responses not validate against the schema. In result I have a JSONPatch files to fix various parts of the service's openapi description.
- Enhancements for programming languages
  - Some OAS elements could be translated to language-level constructs, like `enum`s. Problem is, `enum` in OAS is a list and in programming languages it's typically a kind of a mapping (name := value). So I also have JSON files that add names (x-...) to enum values.
  - Similar case for names: keys in JSON allow much broader set of characters than some programming languages. Character escaping leads to ugly names. So I have another x- attribute to provide friendly aliases.

**Comment by lornajane** [comment]
Some good use cases from @philsturgeon in comment https://github.com/OAI/Overlay-Specification/discussions/45#discussioncomment-10473323

**Comment by mikeschinkel** [comment]
Here is a [discussion post](https://github.com/OAI/Overlay-Specification/discussions/108) about three (3) of the use-cases I have needed to support codegen:

1. Rename a schema and change one of the properties of the renamed schema
2. Delete selected schema properties
3. Subset the OpenAPI doc based on a criteria, in my case the values of the endpoints

**Comment by philsturgeon** [comment]
I've just used Overlays to add Microcks-specific mocking examples to an OpenAPI which would otherwise look a bit weird with them in due to templating syntax showing up in documentation.

https://github.com/bump-sh-examples/train-travel-api/blob/main/.microcks/overlays.yaml
